File: android/src/main/java/expo/modules/tencent_trtc/ExpoTencentTRTCModule.kt
package expo.modules.tencent_trtc

import android.os.Bundle
import androidx.core.os.bundleOf
import com.tencent.liteav.audio.TXAudioEffectManager
import com.tencent.trtc.TRTCCloud
import com.tencent.trtc.TRTCCloudDef
import com.tencent.trtc.TRTCCloudListener
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.URL
import java.util.ArrayList

val notInitializedException =
    CodedException("ERR_NOT_INITIALIZED", "请先调用initTRTCCloud()方法初始化。", null)

class ExpoTencentTRTCModule : Module() {
    var trtcCloud: TRTCCloud? = null

    override fun definition() = ModuleDefinition {
        Name("ExpoTencentTRTC")

        Events("onTRTCEvent")

        OnDestroy {
            TRTCCloud.destroySharedInstance()
        }

        Function("initTRTCCloud") {
            trtcCloud = TRTCCloud.sharedInstance(appContext.reactContext)
            trtcCloud?.addListener(object : TRTCCloudListener() {
                override fun onError(errCode: Int, errMsg: String?, extraInfo: Bundle?) {
                    sendEvent(
                        "onTRTCEvent",
                        bundleOf(
                            "name" to "onError",
                            "code" to errCode,
                            "message" to errMsg,
                            "extraInfo" to extraInfo
                        )
                    )
                }

                override fun onEnterRoom(result: Long) {
                    sendEvent("onTRTCEvent", bundleOf("name" to "onEnterRoom", "result" to result))
                }

                override fun onExitRoom(reason: Int) {
                    sendEvent("onTRTCEvent", bundleOf("name" to "onExitRoom", "reason" to reason))
                }

                override fun onUserVoiceVolume(
                    userVolumes: ArrayList<TRTCCloudDef.TRTCVolumeInfo>?,
                    totalVolume: Int
                ) {
                    val userVolumesData = userVolumes?.map {
                        return@map bundleOf(
                            "pitch" to it.pitch,
                            "spectrumData" to it.spectrumData,
                            "userId" to it.userId,
                            "vad" to it.vad,
                            "volume" to it.volume
                        )
                    }
                    sendEvent(
                        "onTRTCEvent",
                        bundleOf(
                            "name" to "onUserVoiceVolume",
                            "userVolumes" to userVolumesData,
                            "totalVolume" to totalVolume
                        )
                    )
                }

                override fun onNetworkQuality(
                    localQuality: TRTCCloudDef.TRTCQuality?,
                    remoteQuality: ArrayList<TRTCCloudDef.TRTCQuality>?
                ) {
                    val localQualityBundle: Bundle? =
                        if (localQuality != null) bundleOf(
                            "userId" to localQuality.userId,
                            "quality" to localQuality.quality
                        ) else null;
                    val remoteQualityBundle: List<Bundle>? =
                        remoteQuality?.map {
                            bundleOf(
                                "userId" to it.userId,
                                "quality" to it.quality
                            )
                        };
                    sendEvent(
                        "onTRTCEvent",
                        bundleOf(
                            "name" to "onNetworkQuality",
                            "localQuality" to bundleOf(
                                "userId" to localQuality?.userId,
                                "localQuality" to localQualityBundle,
                                "remoteQuality" to remoteQualityBundle
                            )
                        )
                    )
                }
            })
        }

        AsyncFunction("enterRoom") { options: EnterRoomParams ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            val enterRoomParams = TRTCCloudDef.TRTCParams()
            enterRoomParams.sdkAppId = options.sdkAppId
            enterRoomParams.userId = options.userId
            enterRoomParams.roomId = options.roomId ?: 0
            enterRoomParams.strRoomId = options.strRoomId ?: ""
            enterRoomParams.role = options.role.value
            trtcCloud?.enterRoom(enterRoomParams, options.scene.value)
        }

        AsyncFunction("switchRole") { role: Int, privateMapKey: String? ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            if (privateMapKey != null) {
                trtcCloud?.switchRole(role, privateMapKey)
            } else {
                trtcCloud?.switchRole(role)
            }
        }

        AsyncFunction("startLocalAudio") { audioQuality: Int ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            trtcCloud?.startLocalAudio(audioQuality)
        }

        AsyncFunction("setAudioPlaybackVolume") { volume: Int ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            trtcCloud?.audioPlayoutVolume = volume
        }

        AsyncFunction("exitRoom") {
            if (trtcCloud == null) {
                throw notInitializedException
            } else {
                trtcCloud?.exitRoom();
            }
        }

        AsyncFunction("muteLocalAudio") { mute: Boolean ->
            if (trtcCloud == null) {
                throw notInitializedException
            }

            trtcCloud?.muteLocalAudio(mute)
        }

        AsyncFunction("switchRoom") { options: SwitchRoomParams ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            val config = TRTCCloudDef.TRTCSwitchRoomConfig()
            config.roomId = options.roomId ?: 0
            config.strRoomId = options.strRoomId ?: ""
            config.privateMapKey = options.privateMapKey
            config.userSig = options.userSig
            trtcCloud?.switchRoom(config)
        }

        AsyncFunction("muteRemoteAudio") { userId: String, mute: Boolean ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            trtcCloud?.muteRemoteAudio(userId, mute)
        }

        AsyncFunction("muteAllRemoteAudio") { mute: Boolean ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            trtcCloud?.muteAllRemoteAudio(mute)
        }

        AsyncFunction("setRemoteAudioVolume") { userId: String, volumn: Int ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            trtcCloud?.setRemoteAudioVolume(userId, volumn)
        }

        AsyncFunction("enableAudioVolumeEvaluation") { options: AudioVolumeEvaluationOptions ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            val params = TRTCCloudDef.TRTCAudioVolumeEvaluateParams()
            params.interval = options.interval
            params.enableVadDetection = options.enableVadDetection
            params.enablePitchCalculation = options.enablePitchCalculation
            params.enableSpectrumCalculation = options.enableSpectrumCalculation
            trtcCloud?.enableAudioVolumeEvaluation(options.enable, params)

        }

        AsyncFunction("setConsoleLogEnabled") { enable: Boolean ->
            if (trtcCloud == null) {
                throw notInitializedException
            }

            TRTCCloud.setConsoleEnabled(enable)
        }

        AsyncFunction("setConsoleLogLevel") { level: Int ->
            if (trtcCloud == null) {
                throw notInitializedException
            }

            TRTCCloud.setLogLevel(level)
        }

        AsyncFunction("setVoiceReverbType") { reverbType: Int ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            val allValues = TXAudioEffectManager.TXVoiceReverbType.entries.toTypedArray();
            val enumValue = allValues.find { it.nativeValue == reverbType }
            if (enumValue != null) {
                trtcCloud?.audioEffectManager?.setVoiceReverbType(enumValue)
            }
        }

        AsyncFunction("setVoiceChangerType") { changeType: Int ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            val allValues = TXAudioEffectManager.TXVoiceChangerType.entries.toTypedArray();
            val enumValue = allValues.find { it.nativeValue == changeType }
            if (enumValue != null) {
                trtcCloud?.audioEffectManager?.setVoiceChangerType(enumValue)
            }
        }

        AsyncFunction("switchCamera") { useFrontCamera: Boolean ->
            if (trtcCloud == null) {
                throw notInitializedException
            }
            val deviceManager = trtcCloud?.deviceManager
            deviceManager?.switchCamera(useFrontCamera)
        }


        View(ExpoTencentTRTCView::class) {

            AsyncFunction("startLocalPreview") { view: ExpoTencentTRTCView, useFrontCamera: Boolean ->
                trtcCloud?.startLocalPreview(useFrontCamera, view.contentView)
            }

            AsyncFunction("stopLocalPreview") { view: ExpoTencentTRTCView ->
                trtcCloud?.stopLocalPreview()
            }

            AsyncFunction("startRemoteView") { view: ExpoTencentTRTCView,
                                               userId: String,
                                               streamType: VideoStreamType ->
                trtcCloud?.startRemoteView(
                    userId,
                    streamType.value,
                    view.contentView
                )
            }

            AsyncFunction("stopRemoteView") { view: ExpoTencentTRTCView,
                                              userId: String,
                                              streamType: VideoStreamType ->
                trtcCloud?.stopRemoteView(
                    userId,
                    streamType.value
                )
            }

            AsyncFunction("updateRemoteView") { view: ExpoTencentTRTCView,
                                                userId: String,
                                                streamType: VideoStreamType ->
                trtcCloud?.updateRemoteView(
                    userId,
                    streamType.value,
                    view.contentView
                )
            }

        }
    }
}

File: ios/ExpoTencentTRTCModule.swift
import ExpoModulesCore
import TXLiteAVSDK_TRTC

let notInitializedException = Exception(name: "ERR_NOT_INITIALIZED", description: "请先调用initTRTCCloud()方法初始化。")

public class ExpoTencentTRTCModule: Module, TRTCCloudDelegateWrapperDelegate {
    var trtcCloud: TRTCCloud?
    
    public func definition() -> ModuleDefinition {
        Name("ExpoTencentTRTC")
        
        OnDestroy {
            TRTCCloud.destroySharedInstance()
        }
        
        Events("onTRTCEvent")
        
        /// 初始化SDK
        AsyncFunction("initTRTCCloud") {
            self.trtcCloud = TRTCCloud.sharedInstance()
            let delegate = TRTCCloudDelegateWrapper()
            delegate.delegate = self
            
            self.trtcCloud?.addDelegate(delegate)
            
        }
        /// 进入房间，开始拉流
        AsyncFunction("enterRoom") { (options: EnterRoomParams) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            let params = TRTCParams()
            params.sdkAppId = options.sdkAppId
            params.roomId = options.roomId ?? 0
            params.strRoomId = options.strRoomId ?? ""
            params.userId = options.userId
            params.userSig = options.userSig
            params.role = TRTCRoleType(rawValue: options.role.rawValue) ?? .audience
            
            self.trtcCloud?.enterRoom(params, appScene: TRTCAppScene(rawValue: options.scene.rawValue) ?? .voiceChatRoom)
        }
        
        AsyncFunction("switchRole") { (role: TRTCRole, privateMapKey: String?) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            if let role = TRTCRoleType(rawValue: role.rawValue) {
                if privateMapKey != nil {
                    self.trtcCloud?.switch(role, privateMapKey: privateMapKey!)
                } else {
                    self.trtcCloud?.switch(role)
                }
            }
        }
        
        AsyncFunction("startLocalAudio") { (audioQuality: AudioQuality) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            if let quality = TRTCAudioQuality(rawValue: audioQuality.rawValue) {
                self.trtcCloud?.startLocalAudio(quality)
            }
        }
        
        AsyncFunction("exitRoom") {
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            self.trtcCloud?.exitRoom()
        }
        
        AsyncFunction("setAudioPlaybackVolume") { (volume: Int) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            self.trtcCloud?.setAudioPlayoutVolume(volume)
        }
        
        AsyncFunction("muteLocalAudio") { (mute: Bool) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            self.trtcCloud?.muteLocalAudio(mute)
        }
        
        AsyncFunction("switchRoom") { (options: SwitchRoomParams) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            let config = TRTCSwitchRoomConfig()
            config.roomId = options.roomId ?? 0
            config.strRoomId = options.strRoomId ?? ""
            config.userSig = options.userSig
            config.privateMapKey = options.privateMapKey
            self.trtcCloud?.switchRoom(config)
        }
        
        AsyncFunction("muteRemoteAudio") { (userId: String, mute: Bool) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            self.trtcCloud?.muteRemoteAudio(userId, mute: mute)
        }
        
        AsyncFunction("muteAllRemoteAudio") { (mute: Bool) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            self.trtcCloud?.muteAllRemoteAudio(mute)
        }
        
        AsyncFunction("setRemoteAudioVolume") { (userId: String, volumn: Int32) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            self.trtcCloud?.setRemoteAudioVolume(userId, volume: volumn)
        }
        
        
        AsyncFunction("enableAudioVolumeEvaluation") { (options: AudioVolumeEvaluationOptions) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            let params = TRTCAudioVolumeEvaluateParams()
            params.interval = options.interval
            params.enableVadDetection = options.enableVadDetection
            params.enablePitchCalculation = options.enablePitchCalculation
            params.enableSpectrumCalculation = options.enableSpectrumCalculation
            self.trtcCloud?.enableAudioVolumeEvaluation(options.enable, with: params)
        }
        
        AsyncFunction("setConsoleLogEnabled") { (enabled: Bool) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            TRTCCloud.setConsoleEnabled(enabled)
        }
        
        AsyncFunction("setConsoleLogLevel") { (level: Int) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            if let level = TRTCLogLevel(rawValue: level) {
                TRTCCloud.setLogLevel(level)
            }
        }
        
        
        AsyncFunction("setVoiceReverbType") { (reverbType: ReverbType) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            
            if let reverbType = TXVoiceReverbType(rawValue: reverbType.rawValue) {
                self.trtcCloud?.getAudioEffectManager().setVoiceReverbType(reverbType)
            }
        }
        
        AsyncFunction("setVoiceChangerType") { (changeType: VoiceChangerType) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            
            if let changeType = TXVoiceChangeType(rawValue: changeType.rawValue) {
                self.trtcCloud?.getAudioEffectManager().setVoiceChangerType(changeType)
                
            }
        }
        
        AsyncFunction("switchCamera") { (useFrontCamera: Bool) in
            if self.trtcCloud == nil {
                throw notInitializedException
            }
            let deviceManager = trtcCloud?.getDeviceManager()
            deviceManager?.switchCamera(useFrontCamera)
        }
        
        View(ExpoTencentTRTCView.self) {
            
            AsyncFunction("startLocalPreview") { (view: ExpoTencentTRTCView, useFrontCamera: Bool) in
                self.trtcCloud?.startLocalPreview(useFrontCamera, view: view.contentView)
            }
            
            AsyncFunction("stopLocalPreview") { (view: ExpoTencentTRTCView) in
                self.trtcCloud?.stopLocalPreview()
            }
            
            AsyncFunction("startRemoteView") { (view: ExpoTencentTRTCView,
                                                userId: String,
                                                streamType: VideoStreamType) in
                self.trtcCloud?.startRemoteView(userId,
                                                streamType: TRTCVideoStreamType(rawValue: streamType.rawValue) ?? .small,
                                                view: view.contentView)
            }
            
            AsyncFunction("stopRemoteView") { (view: ExpoTencentTRTCView,
                                               userId: String,
                                               streamType: VideoStreamType) in
                self.trtcCloud?.stopRemoteView(userId,
                                               streamType: TRTCVideoStreamType(rawValue: streamType.rawValue) ?? .small)
            }
            
            AsyncFunction("updateRemoteView") { (view: ExpoTencentTRTCView,
                                                 userId: String,
                                                 streamType: VideoStreamType) in
                self.trtcCloud?.updateRemoteView(view.contentView,
                                                 streamType: TRTCVideoStreamType(rawValue: streamType.rawValue) ?? .small,
                                                 forUser: userId)
            }
            
        }
    }
    
    // Delegate
    func onError(_ errCode: TXLiteAVError, errMsg: String?, extInfo: [AnyHashable : Any]?) {
        sendEvent("onTRTCEvent", ["name": "onError",
                                  "code": errCode.rawValue,
                                  "message": errMsg,
                                  "extraInfo": extInfo])
    }
    
    func onEnterRoom(_ result: Int) {
        sendEvent("onTRTCEvent", ["name": "onEnterRoom", "result": result])
    }
    
    func onExitRoom(_ reason: Int) {
        sendEvent("onTRTCEvent", ["name": "onExitRoom", "reason": reason])
    }
    
    func onUserVoiceVolume(_ userVolumes: [TRTCVolumeInfo], totalVolume: Int) {
        let userVolumesMap: [[String: Any?]] = userVolumes.map { volumeInfo in
            return ["pitch": volumeInfo.pitch,
                    "spectrumData": volumeInfo.spectrumData,
                    "userId": volumeInfo.userId,
                    "vad": volumeInfo.vad,
                    "volume": volumeInfo.volume]
        }
        sendEvent("onTRTCEvent", ["name": "onUserVoiceVolume",
                                  "userVolumes": userVolumesMap,
                                  "totalVolume": totalVolume])
    }
    
    func onNetworkQuality(_ localQuality: TRTCQualityInfo, remoteQuality: [TRTCQualityInfo]) {
        let localQualityMap: [String: Any?] = ["userId": localQuality.userId, "quality": localQuality.quality.rawValue]
        let remoteQualityMap: [[String: Any?]] = remoteQuality.map{["userId": $0.userId, "quality": $0.quality.rawValue]}
        
        sendEvent("onTRTCEvent", ["name": "onNetworkQuality",
                                  "localQuality": localQualityMap,
                                  "remoteQuality": remoteQualityMap])
    }
}


@objc protocol TRTCCloudDelegateWrapperDelegate {
    func onError(_ errCode: TXLiteAVError, errMsg: String?, extInfo: [AnyHashable : Any]?)
    func onEnterRoom(_ result: Int)
    func onExitRoom(_ reason: Int)
    func onUserVoiceVolume(_ userVolumes: [TRTCVolumeInfo], totalVolume: Int)
    func onNetworkQuality(_ localQuality: TRTCQualityInfo, remoteQuality: [TRTCQualityInfo])
}

class TRTCCloudDelegateWrapper: NSObject, TRTCCloudDelegate {
    weak var delegate: TRTCCloudDelegateWrapperDelegate?
    
    func onError(_ errCode: TXLiteAVError, errMsg: String?, extInfo: [AnyHashable : Any]?) {
        delegate?.onError(errCode, errMsg: errMsg, extInfo: extInfo)
    }
    
    func onEnterRoom(_ result: Int) {
        delegate?.onEnterRoom(result)
    }
    
    func onExitRoom(_ reason: Int) {
        delegate?.onExitRoom(reason)
    }
    
    func onUserVoiceVolume(_ userVolumes: [TRTCVolumeInfo], totalVolume: Int) {
        delegate?.onUserVoiceVolume(userVolumes, totalVolume: totalVolume)
    }
    
    func onNetworkQuality(_ localQuality: TRTCQualityInfo, remoteQuality: [TRTCQualityInfo]) {
        delegate?.onNetworkQuality(localQuality, remoteQuality: remoteQuality)
    }
}
